src/openforms/payments/models.py

```diff
@@ -36,37 +35,45 @@ def create_for(
             amount=amount,
         )
         # then update with a unique order_id
-        while True:
-            try:
-                with transaction.atomic():
-                    payment.order_id = self.get_next_order_id()
-                    payment.public_order_id = self.create_public_order_id_for(payment)
-                    payment.save(update_fields=("order_id", "public_order_id"))
-                    break
-            except IntegrityError:
-                # race condition on unique order_id
-                continue
-        return payment
 
-    def get_next_order_id(self) -> int:
-        agg = self.aggregate(Max("order_id"))
-        max_order_id = agg["order_id__max"] or ORDER_ID_START
-        return max_order_id + 1
+        with transaction.atomic():
+            payment.public_order_id = self.create_public_order_id_for(payment)
+            payment.save(update_fields=("public_order_id",))
+
+        return payment
 
     @staticmethod
-    def create_public_order_id_for(payment: "SubmissionPayment") -> str:
+    def create_public_order_id_for(
+        payment: SubmissionPayment, pk: int | None = None
+    ) -> str:
+        """Create a public order ID to be sent to the payment provider."""
+
+        # TODO it isn't really clear what the required format/max length is
+        # for payment providers. Ogone seems to allow up to 40 characters or so,
+        # So this might fail at some point.
         config = GlobalConfiguration.get_solo()
-        prefix = config.payment_order_id_prefix
-        prefix = prefix.replace("{year}", str(payment.created.year))
-        order_id = str(payment.order_id).rjust(ORDER_ID_PAD_LENGTH, "0")
-        return prefix + order_id
+        prefix = config.payment_order_id_prefix.format(year=payment.created.year)
+        if prefix:
+            prefix = f"{prefix}/"
+
+        reference = payment.submission.public_registration_reference
+        if not reference and config.wait_for_payment_to_register:
+            from openforms.submissions.public_references import (  # circ. ref
```

Comment:
this codepath appears to not be covered by tests, but I'd reckon that it's pretty important for what will likely be the popular configuration (register after payment)